### PR TITLE
WT-6894 Implement history store cursor next and prev methods

### PR DESCRIPTION
**Pull request by hbokhari**

### CODE REVIEW

File: src/cursor/cur_hs.c [hunk 1]
```diff
@@ -142,6 +142,50 @@ __wt_hs_cursor_search_near(WT_SESSION_IMPL *session, WT_CURSOR *cursor, int *exa
...
+__curhs_next(WT_CURSOR *cursor)
+{
+    WT_CURSOR *file_cursor;
+    WT_CURSOR_HS *hs_cursor;
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+
+    hs_cursor = (WT_CURSOR_HS *)cursor;
+    file_cursor = hs_cursor->file_cursor;
+    CURSOR_API_CALL(cursor, session, next, CUR2BT(file_cursor));
```
Comment (bvpvamsikrishna) [inline src/cursor/cur_hs.c:159]: We have to consider the btree_id, key, timestamp set for hs_cursor as part of set key. We need to return WT_NOTFOUND based on the criteria set (i.e. if only btree_id is set, we need to return WT_NOTFOUND when we are crossing over to a new btree in history store.). For reference you can check __curhs_prev_visible/__curhs_next_visible

File: src/cursor/cur_hs.c [hunk 2]
```diff
@@ -279,7 +323,7 @@ __curhs_prev_visible(WT_SESSION_IMPL *session, WT_CURSOR_HS *hs_cursor)
 
     WT_ERR(__wt_scr_alloc(session, 0, &datastore_key));
 
-    for (; ret == 0; ret = __wt_hs_cursor_prev(session, file_cursor)) {
```
Comment (quchenhao) [inline src/cursor/cur_hs.c:282]: We shouldn't replace this function with std cursor call. We need to use the low level function to ensure we see everything in history store.

Comment (quchenhao) [reply src/cursor/cur_hs.c:282]: The function of prev and next is to call the low level function __wt_hs_cursor_prev  and check the visibility using __curhs_prev_visible and __curhs_next_visible until it finds the next visible record to user.

File: src/cursor/cur_hs.c (hunk 1 above)
Comment (hbokhari) [reply src/cursor/cur_hs.c:162]: Fixed.

**Review by quchenhao — APPROVED** [review]
lgtm

File: src/cursor/cur_hs.c [hunk 3]
```diff
@@ -142,6 +145,52 @@ __wt_hs_cursor_search_near(WT_SESSION_IMPL *session, WT_CURSOR *cursor, int *exa
...
+__curhs_next(WT_CURSOR *cursor)
+{
+    WT_CURSOR *file_cursor;
+    WT_CURSOR_HS *hs_cursor;
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+
+    hs_cursor = (WT_CURSOR_HS *)cursor;
+    file_cursor = hs_cursor->file_cursor;
+    CURSOR_API_CALL(cursor, session, next, CUR2BT(file_cursor));
```
Comment (quchenhao) [inline src/cursor/cur_hs.c:162]: I think I'm using CURSOR_API_CALL_PREPARE_ALLOWED. This allows the session that runs prepared operation also to call this.

Comment (hbokhari) [reply src/cursor/cur_hs.c]: Fixed.

File: src/cursor/cur_hs.c [hunk 4]
```diff
@@ -501,7 +550,7 @@ __curhs_search_near(WT_CURSOR *cursor, int *exactp)
              * be more than one record away the end of our target key/timestamp range. Keep
              * iterating backwards until we land on our key.
              */
-            while ((ret = file_cursor->prev(cursor)) == 0) {
+            while ((ret = __wt_hs_cursor_prev(session, file_cursor)) == 0) {
```
Comment (bvpvamsikrishna) [inline src/cursor/cur_hs.c:553]: input parameter should be hs_cursor right ?

Comment (quchenhao) [reply src/cursor/cur_hs.c:553]: No, we are walking the underlying cursor here.

Comment (bvpvamsikrishna) [reply src/cursor/cur_hs.c:553]: Why caller has to bother about underlying cursor? shouldn't the abstration take care of that?

Comment (quchenhao) [reply src/cursor/cur_hs.c:553]: We are in the implementation of abstraction. I don't quite understand what you mean.

Comment (hbokhari) [reply src/cursor/cur_hs.c:553]: Just to be clear I am just fixing an existing bug and this not directly related to title of the ticket.

**Review by bvpvamsikrishna — APPROVED** [review]

**Pull request merged**